### ISSUE TITLE
Automatizing Strings + inline commented

### ISSUE THREAD

**Issue by jags613** [vivisect/vivisect#123]
Hi!

I see that strings only show up in the strings pane if manually designated as such. 

Is it possible to have automatic analysis to add strings (even if it adds some garbage or false positives)?
And if so, could these then be shown in line when referenced in the assembly/graph views?

This can be very helpful when analyzing for functionality.

Thanks!

**Comment by invisig0th** [comment]
The automatic analysis pass (for most plaforms/formats) should be able to identify string constants in general.  Would you mind sharing a bit more about the circumstances?  Also, vivisect *strongly* prefers "correctness" over "completeness" ( primarily designed to facilitate algorithmic approaches rather than humon review ) so loosening strings and causing more false positives will probably not work.

**Comment by jags613** [comment]
In that case, let me backtrack, the strings pane doesn't include anything (running vivisect on OS10.11) both when analyzing PE or Mach-O binaries. If I find the string and designate it as such then it'll show up but not otherwise.

**Comment by invisig0th** [comment]
Did you create the workspace yourself ( via the API ) or via the cli?  ( If you were working directly with the workspace API, dont forget to call analyze() ).  Do you see functions and such?

**Comment by jags613** [comment]
Not sure where the analyze() instruction would go (haven't found documentation for the API, though would love a means to invoke standalone emulator features if that's already in there). I'm creating/launching with vivbin -B <binary>. Am I missing a step for the analysis?

**Comment by invisig0th** [comment]
Nope, vivbin -B should be doing the analysis pass and then saving the results (including emulation).

**Comment by cmaruti** [comment]
Here you go with a sample snippet of code I did in the past to practice with the API
```python
import sys

VDB_ROOT = "K:\\vivisect_20140803"
sys.path.append(VDB_ROOT)

import vivisect.cli as viv_cli
import vivisect.parsers as viv_parsers

TARGET_NAME = "C:\\Windows\System32\calc.exe"
DEBUG = True

vw = viv_cli.VivCli()
parsemod = viv_parsers.guessFormatFilename(TARGET_NAME)
vw.loadFromFile(TARGET_NAME, fmtname=parsemod)
vw.analyze()

print "Segments #: %d" % len(vw.getSegments())

if DEBUG:
  print "[Segments]"
  for seg in vw.getSegments():
    print seg

print "Imports #: %d" % len(vw.getImports())

if DEBUG:
  print "[Imports]"
  for imp in vw.getImports():
    print imp

print "Exports #: %d" % len(vw.getExports())

if DEBUG:
  print "[Exports]"
  for exp in vw.getExports():
    print exp

print "Functions #: %d" % len(vw.getFunctions())

if DEBUG:
  print "[Functions list]"
  for func in vw.getFunctions():
    print "\t0x%08x %s" % (func, vw.getName(func))


print "Names #: %d" % len(vw.getNames())

if DEBUG:
  print "[Names]"
  for addr, name in vw.getNames():
    print "\t0x%08x %s" % (addr, name)```

**Comment by atlas0fd00m** [comment]
i like to set verbose mode to 4 before calling analyze:

...
vw.verbose=4
vw.analyze()
...

but that's because i love all the nummy details (probably cuz it helps me develop)

On Tuesday, April 12, 2016 08:28:09 AM Cristiano Maruti wrote:
> Here you go with a sample snippet of code I did in the past to practice with the API
> ```python
> import sys
> 
> VDB_ROOT = "K:\\vivisect_20140803"
> sys.path.append(VDB_ROOT)
> 
> import vivisect.cli as viv_cli
> import vivisect.parsers as viv_parsers
> 
> TARGET_NAME = "C:\\Windows\System32\calc.exe"
> DEBUG = True
> 
> vw = viv_cli.VivCli()
> parsemod = viv_parsers.guessFormatFilename(TARGET_NAME)
> vw.loadFromFile(TARGET_NAME, fmtname=parsemod)
> vw.analyze()
> 
> print "Segments #: %d" % len(vw.getSegments())
> 
> if DEBUG:
>   print "[Segments]"
>   for seg in vw.getSegments():
>     print seg
> 
> print "Imports #: %d" % len(vw.getImports())
> 
> if DEBUG:
>   print "[Imports]"
>   for imp in vw.getImports():
>     print imp
> 
> print "Exports #: %d" % len(vw.getExports())
> 
> if DEBUG:
>   print "[Exports]"
>   for exp in vw.getExports():
>     print exp
> 
> print "Functions #: %d" % len(vw.getFunctions())
> 
> if DEBUG:
>   print "[Functions list]"
>   for func in vw.getFunctions():
>     print "\t0x%08x %s" % (func, vw.getName(func))
> 
> 
> print "Names #: %d" % len(vw.getNames())
> 
> if DEBUG:
>   print "[Names]"
>   for addr, name in vw.getNames():
>     print "\t0x%08x %s" % (addr, name)```
> 
> ---
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
-- 
4B5255474B345446454255584749444F4E34514732594C484E465253413544504542525736334C514F5632474B345454465151473633544D5045514732594C484E46525753594C4F4F4D5841553D3D3D
> https://github.com/vivisect/vivisect/issues/123#issuecomment-208962075

**Comment by jags613** [comment]
Thank you for the script! I'll play around with this a bit.

Should I chuck up the no-string problem to OSX issues or is there a way to fix this?

**Comment by atlas0fd00m** [comment]
@invisig0th visi: VivWorkspace.detectString() is horribly broken.  it bails if the second byte of a potential string is not a location....

**Comment by invisig0th** [comment]
Commited your fix.  Without more specifics to address i'd like to close this ticket ( and start fresh if there are more issues )

**Comment by atlas0fd00m** [comment]
works for me.  @jags613 work for you?  please pull the latest code and test.

**Comment by jags613** [comment]
Thank you for the hard work guys! 
Strings work perfectly on OSX host analyzing Windows binaries.
Not getting good results on macho binaries, either no strings at all or some chopped up strings.
